File: OC-LongImage/OC-LongImage/ViewController.m
//
//  ViewController.m
//  OC-LongImage
//
//  Created by jingwei on 2025/7/28.
//

#import "ViewController.h"
#import <SDWebImage/UIImageView+WebCache.h>
#import "UIView+WebCache.h"
#import "FLAnimatedImage.h"
#import "SDWebImageDownloader.h"


@interface CYDynamicImageView : FLAnimatedImageView
@end
@implementation CYDynamicImageView
@end

@interface ViewController ()

@property (nonatomic, strong) CYDynamicImageView *imgView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.imgView = [[CYDynamicImageView alloc]init];
    self.imgView.backgroundColor = UIColor.lightGrayColor;
    self.imgView.contentMode = UIViewContentModeScaleAspectFill;
    self.imgView.frame = CGRectMake(50, 100, 200, 200);
    self.imgView.layer.masksToBounds = YES;
    [self.view addSubview:self.imgView];
    
    //[[SDWebImageManager sharedManager].imageCache clearWithCacheType:SDImageCacheTypeAll completion:nil];
}

/*
 1.直接使用 sd_setImageWithURL 解码占用大量内存 80m
 2.测试 sd_internal 解码占用大量内存 80m
 3.使用 自行解码 ok 占用8m内存
 */
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    NSURL *url = [NSURL URLWithString:@"https://cx-yyz-1251125656.cos.ap-shanghai.myqcloud.com//10_1_8_190/Dynamic/687efe485892fc5a8e8d33d0.jpg"];
//    [self.imgView sd_setImageWithURL:url placeholderImage:nil];

//    [self.imgView sd_internalSetImageWithURL:url placeholderImage:nil options:0 context:nil setImageBlock:^(UIImage * _Nullable image, NSData * _Nullable imageData, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
//    } progress:nil completed:nil];
    
    SDWebImageDownloader *loader = [SDWebImageDownloader sharedDownloader];
    [loader requestImageWithURL:url options:0 context:nil progress:nil completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, BOOL finished) {
        CGFloat size = image.size.width;
        [self setImageWithUndecodedData:data cropRect:CGRectMake(0, 0, image.size.width, image.size.height)];
    }];
}

- (void)setImageWithUndecodedData:(NSData *)imageData cropRect:(CGRect)cropRect {
    if (!imageData) return;
    
    // 后台线程处理，避免阻塞主线程
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
//        @autoreleasepool {
            // 1. 创建图像源（不立即解码）
            CGImageSourceRef imageSource = CGImageSourceCreateWithData((__bridge CFDataRef)imageData, NULL);
            if (!imageSource) return;
        NSLog(@"---- 创建图像源（不立即解码）-----");
            
            // 2. 获取原始图像的像素尺寸
            CFDictionaryRef properties = CGImageSourceCopyPropertiesAtIndex(imageSource, 0, NULL);
            NSNumber *widthNum = (__bridge NSNumber *)CFDictionaryGetValue(properties, kCGImagePropertyPixelWidth);
            NSNumber *heightNum = (__bridge NSNumber *)CFDictionaryGetValue(properties, kCGImagePropertyPixelHeight);
            CGFloat pixelWidth = [widthNum floatValue];
            CGFloat pixelHeight = [heightNum floatValue];
            CFRelease(properties);
            
            // 3. 计算裁剪区域（转换为像素坐标）
            CGRect pixelCropRect = CGRectMake(
                cropRect.origin.x,
                cropRect.origin.y,
                cropRect.size.width,
                cropRect.size.height
            );
            
            // 校验裁剪区域有效性
            if (pixelCropRect.origin.x < 0 || pixelCropRect.origin.y < 0 ||
                pixelCropRect.origin.x + pixelCropRect.size.width > pixelWidth ||
                pixelCropRect.origin.y + pixelCropRect.size.height > pixelHeight) {
                CFRelease(imageSource);
                return;
            }

            
//            // 5. 对完整图像进行裁剪
//            CGImageRef croppedCGImage = CGImageCreateWithImageInRect(fullImage, pixelCropRect);
//            CGImageRelease(fullImage); // 及时释放完整图像，减少内存占用
//            if (!croppedCGImage) return;
            

//            CGImageRelease(croppedCGImage);
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            
            // 4. 兼容方案：先获取完整图像（仅解码一次）
            CGImageRef fullImage = CGImageSourceCreateImageAtIndex(imageSource, 0, NULL);
            CFRelease(imageSource);
            if (!fullImage) return;
            NSLog(@"---- 先获取完整图像（仅解码一次）-----");

            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                // 6. 转换为 UIImage 并在主线程显示
                UIImage *croppedImage = [[UIImage alloc] initWithCGImage:fullImage
                                                                   scale:[UIScreen mainScreen].scale
                                                             orientation:UIImageOrientationUp];
                CGImageRelease(fullImage);
                NSLog(@"---- 转换为 UIImage 并在主线程显示）-----");

                dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                    
                    UIImage *img = [croppedImage imageByPreparingForDisplay];
                    NSLog(@"---- imageByPreparingForDisplay-----");

                    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                        dispatch_async(dispatch_get_main_queue(), ^{
                            self.imgView.image = img;
                        });
                    });
                });
                
                
                
            });
        });


//        }
    });
}
@end

File: Swift-LongImage/Swift-LongImage/ViewController.swift
//
//  ViewController.swift
//  Swift-LongImage
//
//  Created by jingwei on 2025/7/29.
//  参考：https://blog.devlxx.com/2022/02/15/iOS%E5%9B%BE%E7%89%87%E6%98%BE%E7%A4%BA%E5%8E%9F%E7%90%86%E4%B8%8E%E4%BC%98%E5%8C%96%E6%80%9D%E8%B7%AF/

import UIKit
import Kingfisher

class ViewController: UIViewController {

    let img = UIImageView(frame: CGRect(x: 50, y: 100, width: 200, height: 200))

    override func viewDidLoad() {
        super.viewDidLoad()
        img.backgroundColor = .lightGray
        img.contentMode = .scaleAspectFill
        view.addSubview(img)
        KingfisherManager.shared.cache.clearCache()
    }

    // MARK: - 原理就是 只解码需要的部分 不需要的部分全部舍弃
    /*
     解码代码
     1。kCGImageSourceShouldCacheImmediately设置为true，创建缩略图时会直接解码
     2。SDWebImage/YYWebImage 调用 -  CGBitmapContextCreate
        sd - + (CGImageRef)CGImageCreateDecoded:(CGImageRef)cgImage orientation:(CGImagePropertyOrientation)orientation
        yy - YYCGImageCreateDecodedCopy
     3.KingFisher - UIGraphicsBeginImageContextWithOptions - 最终也是调用的CGBitmapContextCreate。
        public func decoded(scale: CGFloat) -> KFCrossPlatformImage {
     */
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        let url = URL(string: "https://cx-yyz-1251125656.cos.ap-shanghai.myqcloud.com/10_1_8_190/Dynamic/6827109415774b5d45ef6432.jpeg")
        
        // 内存占用 73.3
        // img.kf.setImage(with: url)
        
        // 长图使用 // 内存占用 22.5
        // img.kf.setImage(with: url, options: [.processor(GrayscaleImageProcessor(hwRatio: 1))])
        
        // 正常的图可以使用  相当于 对分辨率进行压缩
        let processor = DownsamplingImageProcessor(size: CGSize(width: 800, height: 800))
        img.kf.setImage(with: url, placeholder: nil, options: [.processor(processor)])
    }
    
    
    /*
     另外，之前提到的DownSampling这些第三方网络框架都支持。在SDWebImage框架中，通过context设置SDWebImageContextImageThumbnailPixelSize即可调整采样。
     DownSampling（降低采样）
     */
    /// 原图如果是1000 * 1000 我们写100 * 100 会生成100* 100 的位图
    func downsample(imageAt imageURL: URL, to pointSize: CGSize, scale: CGFloat) -> UIImage {
        let sourceOpt = [kCGImageSourceShouldCache : false] as CFDictionary
        // 其他场景可以用createwithdata (data并未decode,所占内存没那么大),
        let source = CGImageSourceCreateWithURL(imageURL as CFURL, sourceOpt)!
        
        // kCGImageSourceShouldCacheImmediately设置为true，创建缩略图时会直接解码
        let maxDimension = max(pointSize.width, pointSize.height) * scale
        let downsampleOpt = [kCGImageSourceCreateThumbnailFromImageAlways : true,
                                     kCGImageSourceShouldCacheImmediately : true ,
                               kCGImageSourceCreateThumbnailWithTransform : true,
                                      kCGImageSourceThumbnailMaxPixelSize : maxDimension] as CFDictionary
        let downsampleImage = CGImageSourceCreateThumbnailAtIndex(source, 0, downsampleOpt)!
        
        return UIImage(cgImage: downsampleImage)
    }
    
}


struct GrayscaleImageProcessor: ImageProcessor {
    let identifier = "com.example.kingfisher.grayscale"

    public var hwRatio: CGFloat = 1

    public init(hwRatio: CGFloat) {
        self.hwRatio = hwRatio
    }
    
    func process(item: ImageProcessItem, options: KingfisherParsedOptionsInfo) -> KFCrossPlatformImage? {
        switch item {
        case .image(let image):
            return image
        case .data(let data):
            if let image = setImage(withUndecodedData: data, hwRatio: self.hwRatio) {
                return image
            }
            return DefaultImageProcessor.default.process(item: item, options: options)
        @unknown default:
            return nil
        }
    }

    func setImage(withUndecodedData imageData: Data?, hwRatio: CGFloat) -> UIImage? {
        guard let imageData = imageData else { return nil }
        // 1. 创建图像源（不立即解码）
        guard let imageSource = CGImageSourceCreateWithData(imageData as CFData, nil) else { return nil }
        
        // 2. 获取原始图像的像素尺寸
        guard let properties = CGImageSourceCopyPropertiesAtIndex(imageSource, 0, nil) as? [CFString: Any],
              let pixelWidth = properties[kCGImagePropertyPixelWidth] as? CGFloat,
              let pixelHeight = properties[kCGImagePropertyPixelHeight] as? CGFloat else {
            return nil
        }
        let cropRect = CGRect(x: 0, y: 0, width: pixelWidth, height: pixelWidth * hwRatio)
        
        // 4. 创建一个未解码的 CGImage
        guard let fullImage = CGImageSourceCreateImageAtIndex(imageSource, 0, nil) else { return nil }
        guard let croppedCGImage = fullImage.cropping(to: cropRect) else {
            return nil
        }
        let croppedImage = UIImage(cgImage: croppedCGImage, scale: UIScreen.main.scale, orientation: .up)
        return croppedImage
    }
}

/*
 class CYDynamicBaseImageView: AnimatedImageView {
     func setImageFid(_ fid: String?, sourceSize: CGSize?, showSize: CGSize) {
         guard let fid = fid else { return }
         let url = URL(string: imgUrl(fid))
         var hwRadio: CGFloat = 1
         if let sourceSize = sourceSize {
             hwRadio = sourceSize.height / sourceSize.width
         }
         let placeholder = UIImage(named: "dynamic_placeholder")
         if hwRadio > 3 || hwRadio < 0.33 {
             self.kf.setImage(with: url, placeholder: placeholder, options: [.processor(GrayscaleImageProcessor(showSize: showSize))])
         } else {
             if fid.lowercased().hasSuffix(".gif") {
                 self.kf.setImage(with: url, placeholder: placeholder)
             } else {
                 let processor = DownsamplingImageProcessor(size: CGSize(width: 800, height: 800))
                 self.kf.setImage(with: url, placeholder: placeholder, options: [.processor(processor)])
             }
         }
     }
     
     func imgUrl(_ fid: String) -> String {
         guard let cos = CYLoginInfoManager.share().serListModel?.cosUrl else { return fid }
         return cos + "/" + fid
     }
 }
 
 // 超长图/超宽图 显示
 struct GrayscaleImageProcessor: ImageProcessor {
     let identifier = "com.example.kingfisher.sizedecode"

     public var showSize: CGSize = CGSizeMake(1, 1)

     public init(showSize: CGSize) {
         self.showSize = showSize
     }
     
     func process(item: ImageProcessItem, options: KingfisherParsedOptionsInfo) -> KFCrossPlatformImage? {
         switch item {
         case .image(let image):
             return image
         case .data(let data):
             if let image = setImage(withUndecodedData: data, showSize: self.showSize) {
                 return image
             }
             return DefaultImageProcessor.default.process(item: item, options: options)
         @unknown default:
             return nil
         }
     }

     func setImage(withUndecodedData imageData: Data?, showSize: CGSize) -> UIImage? {
         guard let imageData = imageData else { return nil }
         guard let imageSource = CGImageSourceCreateWithData(imageData as CFData, nil) else { return nil }
         guard let properties = CGImageSourceCopyPropertiesAtIndex(imageSource, 0, nil) as? [CFString: Any],
               let pixelWidth = properties[kCGImagePropertyPixelWidth] as? CGFloat,
               let pixelHeight = properties[kCGImagePropertyPixelHeight] as? CGFloat else {
             return nil
         }
         // 计算与显示区域等比例的裁剪尺寸
         let aspectRatio = showSize.width / showSize.height
         var cropWidth = pixelWidth
         var cropHeight = pixelHeight

         // 根据宽高比调整裁剪尺寸，保持显示比例
         if aspectRatio > 1 {
             // 超宽图：以高度为基准计算宽度
             cropWidth = min(cropHeight * aspectRatio, pixelWidth)
         } else {
             // 超长图/正方形：以宽度为基准计算高度
             cropHeight = min(cropWidth / aspectRatio, pixelHeight)
         }

         let cropRect = CGRect(x: 0, y: 0, width: cropWidth, height: cropHeight)
         guard let fullImage = CGImageSourceCreateImageAtIndex(imageSource, 0, nil) else { return nil }
         guard let croppedCGImage = fullImage.cropping(to: cropRect) else {
             return nil
         }
         let croppedImage = UIImage(cgImage: croppedCGImage, scale: UIScreen.main.scale, orientation: .up)
         return croppedImage
     }
 }

*/
